packages/@aws-cdk/aws-ecs/lib/ec2/ec2-service.ts

```diff
@@ -198,6 +199,50 @@ export class Ec2Service extends BaseService implements elb.ILoadBalancerTarget {
     });
   }
 
+  /**
+   * Enable CloudMap service discovery for the service
+   */
+  public enableServiceDiscovery(props: ServiceDiscoveryProps): cloudmap.Service {
+    // Determine DNS type based on network mode
+    const networkMode = this.taskDefinition.networkMode;
+
+    if (networkMode === NetworkMode.None) {
+      throw new Error("Cannot use a service discovery if NetworkMode is None. Use Host or AwsVpc instead.");
+    }
+    // Bridge or host network mode requires SRV records
+    let dnsType = props.dnsType;
+
+    if (networkMode === NetworkMode.Bridge || networkMode === NetworkMode.Host) {
+      if (dnsType !== cloudmap.DnsType.Srv) {
+        throw new Error("Cannot use a service discovery if NetworkMode is None. Use Host or AwsVpc instead.");
+      }
+      dnsType = cloudmap.DnsType.Srv;
+    }
+
+    if (networkMode === NetworkMode.AwsVpc && props.dnsType === undefined ) {
+      // ECS does not currently support IPv6, so AAAA records not possible
+      dnsType = cloudmap.DnsType.A;
+    }
+
+    const cloudmapService = new cloudmap.Service(this, 'CloudmapService', {
+      namespaceId: props.namespaceId, // FIXME use INamespace instead -- validate if cluster has called addNamespace set?
+      name: props.name,
+      dnsType: dnsType!,
+      healthCheckCustomConfig: { failureThreshold: props.failureThreshold || 1 }
+    });
+
+    const serviceArn = cloudmapService.serviceArn;
+
+    // add Cloudmap service to the ECS Service's serviceRegistry
+    this.serviceRegistries.push({
+      registryArn: serviceArn,
+      containerName: this.taskDefinition.defaultContainer!.node.id,
```

Comment:
Check for existence of a default container. It might not exist if `addContainer()` hasn't been called yet.